### PR TITLE
Modifications in relation to expiring tokens

### PR DESCRIPTION
**Pull request by malek0512**
when a token is expired, we delete it from the local storage and ask a new one to the api

### CODE REVIEW

**Comment by lchabert** [comment]
Hello,

What is the criteria for deleting access token. As i can see in your pull request, it's only based on response =! 200.
I think you should be more specifique, return a special code, and match against it.

Regards.

**Comment by moimael** [comment]
Good point, according to the RFC, the server must issue a 400 Bad Request with an ascii error code containing "invalid_grant" in case of token expiration.

```

5.2.  Error Response

   The authorization server responds with an HTTP 400 (Bad Request)
   status code (unless specified otherwise) and includes the following
   parameters with the response:

   error
         REQUIRED.  A single ASCII [USASCII] error code from the
         following:

         invalid_request
               The request is missing a required parameter, includes an
               unsupported parameter value (other than grant type),
               repeats a parameter, includes multiple credentials,
               utilizes more than one mechanism for authenticating the
               client, or is otherwise malformed.

         invalid_client
               Client authentication failed (e.g., unknown client, no
               client authentication included, or unsupported
               authentication method).  The authorization server MAY
               return an HTTP 401 (Unauthorized) status code to indicate
               which HTTP authentication schemes are supported.  If the
               client attempted to authenticate via the "Authorization"
               request header field, the authorization server MUST
               respond with an HTTP 401 (Unauthorized) status code and
               include the "WWW-Authenticate" response header field
               matching the authentication scheme used by the client.

         invalid_grant
               The provided authorization grant (e.g., authorization
               code, resource owner credentials) or refresh token is
               invalid, expired, revoked, does not match the redirection
               URI used in the authorization request, or was issued to
               another client.

         unauthorized_client
               The authenticated client is not authorized to use this
               authorization grant type.

         unsupported_grant_type
               The authorization grant type is not supported by the
               authorization server.
```

**Comment by malek0512** [comment]
As the [**RFC6750**](https://tools.ietf.org/html/rfc6750#section-3.1) specifies it, the API returns a 401 Unauthorized code error with the specific message _"invalid_token"_ in the _www-authenticate_ field. 
Unfortunately those headers are not accessible from _xmlhttprequest_ object (I might have found a reason to that on [stackoverflow](http://stackoverflow.com/questions/7462968/restrictions-of-xmlhttprequests-getresponseheader?answertab=active#tab-top))

Anyway, the code status 401 and the message _"Unauthorized"_ remains accessible from the xmlhttprequest object. That might not be sufficient to distinguish between specific errors like _"invalid_request"/"invalid_token"/"insufficient_scope"_, but should be enough for assuming the token expired and giving it a second try.

Surely I can test on 401 code status before considering the token expired.
Is that ok ?

Regards

**Comment by lchabert** [comment]
Yes i'm agree, ressource server must return a 401 error code.

But "Invalid_token" error description is not in headers but in body response. Chrome does not allow you to get body content ?

Regards.

**Comment by malek0512** [comment]
You might be right, but from my tests, the body only contains _"Unauthorized"_ message.
But the header contains : _WWW-Authenticate → Bearer realm="Users", error="invalid_token", error_description="expired_token"_

And I can't access to the _WWW-Authenticate_ field with _xmlhttprequest_ object, because some kind of filter (applied by the browser, according to the stackoverflow link above) hides them away.

Regards

**Comment by moimael** [comment]
According to your stackoverflow link it is CORS related, the server must specify `Access-Control-Expose-Headers: "header_name"`, "header_name" being WWW-Authenticate in our case I guess.

Webkit does implement it now: https://bugs.webkit.org/show_bug.cgi?id=41210

So this should work with a little bit of server conf.

**Comment by lchabert** [comment]
According @moimael comment, you should work on server side to add a body response with appropriate message.

File: index.js [hunk 1]
```diff
@@ -27,6 +27,9 @@
  *
  */
 
+// var base_url = 'http://localhost:3000';
```
Comment (moimael) [inline index.js]: This comment should be removed

File: index.js [hunk 2]
```diff
@@ -46,10 +49,10 @@ var gh = (function() {
         chrome.storage.sync.get({
           access_token: null,
         }, function(items) {
-          access_token = items.access_token;
+          access_token = items.access_token;
```
Comment (moimael) [inline index.js]: Extra space should be removed

File: index.js [hunk 3]
```diff
@@ -46,10 +49,10 @@ var gh = (function() {
         chrome.storage.sync.get({
           access_token: null,
         }, function(items) {
-          access_token = items.access_token;
+          access_token = items.access_token; 
           options = {
             'interactive': interactive,
-            url:'https://api.voxity.fr/api/v1/dialog/authorize?client_id=' + clientID +
+            url: base_url+'/api/v1/dialog/authorize?client_id=' + clientID +
```
Comment (moimael) [inline index.js]: You should add spacing around +

File: index.js [hunk 4]
```diff
@@ -106,8 +109,10 @@ var gh = (function() {
       },
 
       removeCachedToken: function(token_to_remove) {
-        if (access_token == token_to_remove)
+        if (access_token == token_to_remove) {
```
Comment (moimael) [inline index.js]: You should use === to do the comparison

File: index.js [hunk 5]
```diff
@@ -161,7 +166,7 @@ var gh = (function() {
     },
     makeCall: function (exten) {
       var message = {
-        action: 'https://api.voxity.fr/api/v1/channel',
+        action: base_url+'/api/v1/channel',
```
Comment (moimael) [inline index.js]: You should add spacing around +

File: index.js (hunk 3 above)
Comment (malek0512) [inline index.js]: Is there any reason to add spacing or removing extra space ? Sorry for asking I just want learn from my faults

Comment (moimael) [inline index.js]: @malek0512 Just good practices to be JSHint/JSLint compliant. Makes the code more readable.

**Comment by malek0512** [comment]
To respond to @lolostates comment _you should work on server side to add a body response with appropriate message._. 
I'm not sure this the best solution for one reason :  _passport-http-bearer_ module used from server side respect RFC specifications. And patching it wouldn't feel right because the module wouldn't respect RFC anymore.

File: index.js (hunk 3 above)
Comment (malek0512) [inline index.js]: ok thanks for the tip. I've rectified it

**Pull request merged**